alns/select/MABSelector.py

```diff
@@ -0,0 +1,216 @@
+        repair operators. Entry (i, j) is True if destroy operator i can be
+        used together with repair operator j, and False otherwise.
+    context_extractor
+        Optional function that takes a State object and returns a context
+        vector for that state that can be passed to a contextual mabwiser
+        bandit. If the MAB algorithm supports it, this context will be used to
+        help predict the next (destroy, repair) combination.
+
+    References
+    ----------
+    .. [1] Emily Strong, Bernard Kleynhans, & Serdar Kadioglu (2021).
+           MABWiser: Parallelizable Contextual Multi-armed Bandits.
+           Int. J. Artif. Intell. Tools, 30(4), 2150021:1–2150021:19.
+    """
+
+    @staticmethod
+    def make_arms(
+        num_destroy: int,
+        num_repair: int,
+        op_coupling: Optional[np.ndarray] = None,
+    ) -> List[str]:
+        """
+        Generates a list of arms for the MAB passed to MABSelector.
+
+        Any MABs passed to MABSelector must be generated with this function,
+        and with the same ``num_destroy``, ``num_repair``, and ``op_coupling``
+        parameters.
+        """
+        if num_destroy < 1 or num_repair < 1:
+            raise ValueError(
+                "num_destroy and num_repair must be greater than 0."
+            )
+        if op_coupling is not None and op_coupling.sum() == 0:
+            raise ValueError("op_coupling must have at least one True entry.")
+
+        # the set of valid operator pairs is equal to the cartesian product
+        # of destroy and repair operators, except we leave out any pairs
+        # disallowed by op_coupling
+        return [
+            f"{d_idx}_{r_idx}"
+            for d_idx, r_idx in itertools.product(
+                range(num_destroy), range(num_repair)
+            )
+            if op_coupling is None or op_coupling[d_idx, r_idx]
+        ]
+
+    def __init__(
+        self,
+        scores: List[float],
+        mab: MAB,
+        num_destroy: int,
+        num_repair: int,
+        op_coupling: Optional[np.ndarray] = None,
+        context_extractor: Optional[
+            Callable[
+                [State], Union[List[Num], np.ndarray, pd.Series, pd.DataFrame]
+            ]
+        ] = None,
+    ):
+        super().__init__(num_destroy, num_repair, op_coupling)
+
+        if any(score < 0 for score in scores):
+            raise ValueError("Negative scores are not understood.")
+
+        if len(scores) < 4:
+            # More than four is OK because we only use the first four.
+            raise ValueError(f"Expected four scores, found {len(scores)}")
+
+        if mab.arms != self.make_arms(num_destroy, num_repair, op_coupling):
+            raise ValueError(
+                "Arms in MAB passed to MABSelector are incorrect. Make sure"
+                + "to generate arms with ``MABSelector.make_arms``"
+            )
+
+        self._mab = mab
+        self._scores = scores
+        self._primed = False
```

Comment:
not sure about "_primed".. do you mean "is_trained"  OR even "is_initial_fit" as in how mabwiser uses it https://github.com/fidelity/mabwiser/blob/master/mabwiser/mab.py#L1160

IF mabwiser has returned that information, I guess we could have used it here --rather than tracking again